### PR TITLE
add vira-dropdown to vira package

### PR DESCRIPTION
**Pull request by electrovir**
# Changes

- add `vira-dropdown` and `vira-dropdown-item`
- add element-book pages for each

# How to test

Unfortunately I do not have deploy previous setup for this so you'll have to test locally:

1. clone this repo locally
2. checkout this branch
3. run `npm ci`
4. run `cd packages/vira-book`
5. run `npm start`
6. test out the `vira-dropdown` element in the vira-book frontend

### CODE REVIEW

File: cspell.config.cjs [hunk 1]
```diff
@@ -8,5 +8,6 @@ module.exports = {
     words: [
         ...baseConfig.words,
         'observavir',
+        'listbox',
```
Comment (sarangspadalkar) [inline cspell.config.cjs:11]: Test comment

**Review by KylerHansen — APPROVED** [review]

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import {ViraDropdownItem} from 'vira';
+import {dropdownPage} from './vira-dropdown.book';
+
+const examples: ReadonlyArray<{
+    title: string;
+    inputs?: typeof ViraDropdownItem.inputsType;
+    customStyle?: CSSResult;
+    customTemplate?: HTMLTemplateResult;
+}> = [
+    {
```
Comment (sarangspadalkar) [inline packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:12]: Please add examples: 
- having not enough horizontal space
- being stretched to take up more space than necessary

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import {BookPageControlTypeEnum, defineBookPage, definePageControl} from 'element-book';
+import {CSSResult, HTMLTemplateResult, html} from 'element-vir';
+import {ViraDropdownItem} from 'vira';
+import {dropdownPage} from './vira-dropdown.book';
+
+const examples: ReadonlyArray<{
```
Comment (sarangspadalkar) [inline packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:6]: Please add examples: 
- having not enough horizontal space
- being stretched to take up more space than necessary

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
+import {BookPageControlTypeEnum, defineBookPage, definePageControl} from 'element-book';
```
Comment (sarangspadalkar) [inline packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:1]: Is this book deployed somewhere? Netlify or github-pages?

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown.element.book.ts [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                ...exampleDropdownOptions,
+                {
+                    id: 42,
+                    label: 'this is disabled',
+                    disabled: true,
+                },
+            ],
+        },
+    },
+    // {
```
Comment (sarangspadalkar) [inline packages/vira-book/src/element-book/entries/dropdown/vira-dropdown.element.book.ts]: This is your personal repo, so I dont know what conventions you have enforced. But I would avoid having commented code.

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import {CSSResult, HTMLTemplateResult, html} from 'element-vir';
+import {ViraDropdownItem} from 'vira';
+import {dropdownPage} from './vira-dropdown.book';
+
+const examples: ReadonlyArray<{
+    title: string;
+    inputs?: typeof ViraDropdownItem.inputsType;
+    customStyle?: CSSResult;
```
Comment (sarangspadalkar) [inline packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:9]: this is not used anywhere.

```suggestion
```

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+        z_debug_forceOpenState: boolean;
+    }>
+>()({
+    tagName: 'vira-dropdown',
+    hostClasses: {
+        'vira-dropdown-disabled': ({inputs}) => !!inputs.isDisabled,
+    },
+    /** Default colors meet APCA contrast for small body text: https://www.myndex.com/APCA/ */
+    cssVars: {
+        'vira-dropdown-trigger-background-color': 'white',
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: These vars values are very generic which can be used by other elements as well. So add them to a shared css file which can be shared between multiple elements.

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+
+        .open-upwards .select-options {
+            border-bottom-left-radius: 0;
+            border-bottom-right-radius: 0;
+            ${viraShadows.menuShadowReversed}
+        }
+
+        ${hostClasses['vira-dropdown-disabled'].selector} {
+            ${viraDisabledStyles}
+
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: remove this space

```suggestion
```

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+            triggerPopUp({open: false, emitEvent: false});
+        } else if (inputs.z_debug_forceOpenState != undefined) {
+            if (!inputs.z_debug_forceOpenState && state.showPopUpResult) {
+                triggerPopUp({emitEvent: false, open: false});
+            } else if (inputs.z_debug_forceOpenState && !state.showPopUpResult) {
+                triggerPopUp({emitEvent: false, open: true});
+            }
+        }
+
+        const selectedOptions: ReadonlyArray<Readonly<ViraDropdownOption>> = inputs.selected.length
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: I think this should be moved inside a function and that function should have unit tests written for it.

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 10]
```diff
@@ -0,0 +1,511 @@
+import {PartialAndUndefined, isLengthAtLeast, joinWithFinalConjunction} from '@augment-vir/common';
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: please add tests for this element.

**Review by sarangspadalkar — CHANGES_REQUESTED** [review]

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts (hunk 4 above)
Comment (electrovir) [reply packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:1]: GitHub Pages: http://electrovir.github.io/vira/book

Thus there is no deploy preview 😔

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown.element.book.ts (hunk 5 above)
Comment (electrovir) [reply packages/vira-book/src/element-book/entries/dropdown/vira-dropdown.element.book.ts]: good catch!

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 7 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: Agreed, that is planned for the future in `theme-vir` but it's not ready yet.

File: cspell.config.cjs
Comment (electrovir) [inline cspell.config.cjs:1]: todo: prevent opening multiple dropdowns at once

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts (hunk 3 above)
Comment (electrovir) [reply packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:6]: done

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown.element.book.ts (hunk 5 above)
Comment (electrovir) [reply packages/vira-book/src/element-book/entries/dropdown/vira-dropdown.element.book.ts]: fixed

File: packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts (hunk 6 above)
Comment (electrovir) [reply packages/vira-book/src/element-book/entries/dropdown/vira-dropdown-item.element.book.ts:9]: Now it is :)

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 8 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: done

File: cspell.config.cjs
Comment (electrovir) [inline cspell.config.cjs:1]: todo: add placeholder input

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 9 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: done

File: cspell.config.cjs
Comment (electrovir) [reply cspell.config.cjs:1]: fixed

Comment (electrovir) [reply cspell.config.cjs:1]: done

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 10 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: done

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 11]
```diff
@@ -0,0 +1,561 @@
...
+    NavSelectEvent,
+    PopUpManager,
+    ShowPopUpResult,
+} from '../util/pop-up-manager';
+import {defineViraElement} from './define-vira-element';
+import {ViraDropdownItem, ViraDropdownOption} from './vira-dropdown-item.element';
+import {ViraIcon} from './vira-icon.element';
+
+function assertUniqueIdProps(options: ReadonlyArray<Readonly<{id: PropertyKey}>>) {
+    const usedIds: Record<PropertyKey, boolean> = {};
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: nit: Generally [Sets](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) are used to store unique values.

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+        /**
+         * If false, this will behave like a single select dropdown, otherwise you can select
+         * multiple.
+         */
+        isMultiSelect: boolean;
+        icon: ViraIconSvg;
+        selectionPrefix: string;
+        isDisabled: boolean;
+        /** For debugging purposes only. Very bad for actual production code use. */
+        z_debug_forceOpenState: boolean;
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: question: what is the thinking process while behing the name of this property? Is there a reason its strats with `z`?

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 13]
```diff
@@ -0,0 +1,561 @@
...
+
+        'vira-dropdown-text-color': '#000000',
+        'vira-dropdown-background-color': 'white',
+        'vira-dropdown-focus-border-color': viraFocusCssVars['vira-focus-outline-color'].default,
+        'vira-dropdown-border-color': '#cccccc',
+        'vira-dropdown-item-hover-background-color': '#d2eaff',
+        'vira-dropdown-item-hover-text-color': '#000000',
+    },
+    styles: ({hostClasses, cssVars}) => css`
+        :host {
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: this styles object is huge, can it be refactored somehow?

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+        selectedChange: defineElementEvent<PropertyKey[]>(),
+        openChange: defineElementEvent<boolean>(),
+    },
+    stateInitStatic: {
+        /** `undefined` means the pop up is not currently showing. */
+        showPopUpResult: undefined as ShowPopUpResult | undefined,
+        popUpManager: perInstance(() => new PopUpManager()),
+        navController: undefined as NavController | undefined,
+    },
+    cleanupCallback({state, updateState}) {
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: when is this callback executed?

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 15]
```diff
@@ -0,0 +1,561 @@
...
+        updateState({showPopUpResult: undefined});
+        state.popUpManager.destroy();
+    },
+    initCallback({state, updateState, host, inputs, dispatch, events}) {
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: when is this callback executed?

File: packages/vira/src/elements/vira-dropdown.element.ts [hunk 16]
```diff
@@ -0,0 +1,561 @@
+import {PartialAndUndefined, joinWithFinalConjunction} from '@augment-vir/common';
```
Comment (sarangspadalkar) [inline packages/vira/src/elements/vira-dropdown.element.ts]: this file is 500 line long, consider moving these utils functions out of this file.

**Review by sarangspadalkar — APPROVED** [review]

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 12 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: Starting with `_` puts it at the top of IntelliSense's auto complete list, unfortunately. `z` puts it at the bottom 😆

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 15 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts:312]: `initCallback` is called whenever an element is connected to the DOM. It's triggered by the browser itself.

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 14 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: https://www.npmjs.com/package/element-vir#other-callbacks

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 15 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: https://www.npmjs.com/package/element-vir#other-callbacks

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 11 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: good point

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 13 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: Maybe some of the children can be extracted into separate elements, I'll take a closer look.

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 16 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: good idea

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 11 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: fixed

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 16 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: functions moved

File: packages/vira/src/elements/vira-dropdown.element.ts (hunk 13 above)
Comment (electrovir) [reply packages/vira/src/elements/vira-dropdown.element.ts]: I split out `options-wrapper` into a separate element. That cut down on the file and styles size a bit.

**Pull request merged**